### ISSUE TITLE
Graph: Inter-cycle dependencies not correctly checked

### ISSUE THREAD

**Issue by wxtim** [cylc/cylc-flow#6523]
@dpmatthews spotted

### Description

graph

```cylc
foo[-P1]:finish => foo
```

... ought to fail validation, because finish implies `foo[-P1]:succeeded? | foo[-P1]:failed?`, so `foo` (implicit `:succeeded`) should also be optional.

I discovered the simpler case `foo[-P1]:succeeded | foo[-P1]:failed => foo` doesn't appear to work either.

### Reproducible Example

This diff will produce failing unit tests (The empty second string is to test what the error message should be, but no error is produced):

```diff
diff --git a/tests/unit/test_graph_parser.py b/tests/unit/test_graph_parser.py
index 75a0bf95a..a38373a14 100644
--- a/tests/unit/test_graph_parser.py
+++ b/tests/unit/test_graph_parser.py
@@ -139,6 +139,16 @@ def test_graph_syntax_errors_2(seq, graph, expected_err):
             'Invalid cycle point offsets only on right',
             id='no-cycle-point-RHS'
         ),
+        param(
+            "foo[-P1]:succeeded | foo[-P1]:failed => foo",
+            '',
+            id='succeed-or-failed-mustbe-optional'
+        ),
+        param(
+            "foo[-P1]:finish => foo?",
+            '',
+            id='lhs-finish-implies-success-optional'
+        ),
     ]
 )
 def test_graph_syntax_errors(graph, expected_err):
```

[diff.txt](https://github.com/user-attachments/files/18092567/diff.txt)

### Expected Behaviour

These graphs should fail validation

**Comment by oliver-sanders** [comment]
Sadly, fixing this will cause some workflows to fail validation which passed before which really isn't great! We've got a large tranche of Cylc 8 workflow migration coming very soon so getting this out of the way quickly would be desirable for us.

However, it's not suitable for a bugfix release, we should put this into a minor release as a breaking change. I have optimistically tagged against 8.4.0 hoping it's a trivial fix which it may well be. If we can't make this milestone without delaying the release, then bump to 8.5.0?

**Comment by wxtim** [comment]
Now up to 4 tests:

```python
        # See https://github.com/cylc/cylc-flow/issues/6523
        # For the next 4 tests:
        param(
            # Yes I know it's circular, but it's here to
            # demonstrate that the test below is broken:
            "foo:finished => foo",
            'Output foo:succeeded can\'t be both required and'
            ' optional',
            id='finish-implies-success-optional'
        ),
        param(
            "foo[-P1]:finish => foo",
            'Output foo:succeeded can\'t be both required and'
            ' optional',
            id='finish-implies-success-optional-offset'
        ),
        param(
            "foo[-P1]:succeeded | foo[-P1]:failed => bar",
            (
                'Opposite outputs foo:succeeded and foo:failed'
                'must both be optional if both are used'
            ),
            id='succeed-or-failed-mustbe-optional'
        ),
        param(
            "foo[-P1]:succeeded? | foo[-P1]:failed? => foo",
            'Output foo:succeeded can\'t be both required and'
            ' optional',
            id='succeed-or-failed-implies-success-optional'
        ),
```

**Comment by hjoliver** [comment]
It's only with inter-cycle offsets, I take it?

**Comment by wxtim** [comment]
> It's only with inter-cycle offsets, I take it?

Yes. 

I think the error is here https://github.com/cylc/cylc-flow/blob/d21c9e534959cb28353dc6865c3f396e899430dd/cylc/flow/graph_parser.py#L543

**Comment by hjoliver** [comment]
@wxtim - hope you don't mind - I took a look at this because it's high-priority and you haven't posted a fix yet.

Initially it was confusing as hell, but I think the problem is just that we only set and check output optionality on the RHS of graph trigger pairs, but we bail out of processing RHS nodes if they have a cycle point offset on grounds that all necessary processing can be done when they appear on the LHS of a pair. That's just below the line you identified above. Once I removed the bail-out, I just had to change to the RHS node regex to allow for offsets.

(Now let's see if the tests pass...)
